wxListCtrl sort indicator fixes

vadz opened the pull request:
@MaartenBent The most important change here is the one enabling the use of sort indicators for virtual list controls, as this looks like an important use case and I don't understand at all why were they disabled in it -- could you please let me know if I'm missing something?

I've also made `ShowSortIndicator()` enable the indicators as otherwise it just silently does nothing, which can't be the right thing to do.

But there are still more problems here: the generic version doesn't handle clicking on the columns automatically, even with sort indicators enabled, while the MSW one does. We need to harmonize this, but I'd actually prefer to remove the automatic update, as it seems simpler to call `ShowSortIndicator()` from the `wxEVT_LIST_COL_CLICK` handler you're going to have to define anyhow than to rely on this happening from inside wx. I.e. I'd rather remove `EnableSortIndicator()` completely and just show the sort indicator if `ShowSortIndicator()` is called and never update it automatically. This seems much simpler to understand than the current situation. What do you think?

vadz committed:
Allow using column sort indicators in virtual wxListCtrl too

There doesn't seem to be any reason to not show sort indicators in the
columns of virtual list controls, so simply remove the test for
IsVirtual() in DrawSortArrow().

This allows d8ec0aa001 (Support sort indicators in wxListCtrl header,
2021-11-28) to work for wxListCtrl with wxLC_VIRTUAL too.

vadz committed:
Remove top level "const" from {Enable,Show}SortIndicator()

Don't use "const int" or "const bool" for parameter types, the "const"
here is ignored and using it is inconsistent with all the rest of the
library.

No real changes.

vadz committed:
Optimize changing sort indicator in wxGenericListCtrl

Don't do anything at all if nothing changes and if the indicator does
change, refresh only the header window and not the whole list control,
which seems unnecessary.

vadz committed:
Let wxListCtrl::ShowSortIndicator() implicitly enable indicators

It doesn't seem right for ShowSortIndicator() to silently do nothing if
EnableSortIndicator() hadn't been called before, so make it enable the
sort indicators if they hadn't been enabled yet.

The alternative would be to assert in this function, but this seems less
useful.

Also add some comments to wxMSW version.

vadz commented:
clang CI failure is bogus

MaartenBent commented:
I'm looking back in my reflog but can't really find why I disabled it for virtual list controls. I must have misread something, because it is also not in my [original branch](https://github.com/MaartenBent/wxWidgets/commit/0030b73751a0f58f08711f0eb402b4c50297248a).

But the generic version works fine for me. Both on Windows and Linux (GTK3). In the listctrl sample I click on the column and the arrow is shown, and when clicking more times, it toggles direction.

I am fine with changing and simplifying the API and do everything in the event handler of the user. It would remove most of the functions, since you'd only need `ShowSortIndicator` (and `RemoveSortIndicator` that just calls `ShowSortIndicator(-1)`).

MaartenBent commented:
Ok, I do notice sometimes an issue with the arrow not always updating in the generic version, because the header is not refresh after clicking it. Adding a `Refresh()` call after toggling `m_sortAsc` in `OnMouse()` seems to fix it. Though this won't be needed anymore with above suggestion.

vadz committed:
Remove wxListCtrl::EnableSortIndicator()

The old API seems unnecessarily complex, it is simpler to just let the
application call ShowSortIndicator() itself from its
wxEVT_LIST_COL_CLICK handler, which needs to be defined anyhow in order
to actually sort the items, rather than require it to enable sort
indicator, explicitly set it initially and then remember to not set it
any longer in response to the column clicks.

Also make RemoveSortIndicator() non-virtual and implement it simply as
ShowSortIndicator(-1) because this actually simplifies the code too.

vadz commented:
I've implemented the proposed changes in the latest commit, please have a look at them if you can. TIA!

MaartenBent commented:
I'll have a detailed look later, but I think you want to remove this too: https://github.com/wxWidgets/wxWidgets/blob/8d1cf0a5019bf03fef2e4549d81fa768d115ac7d/src/generic/listctrl.cpp#L1337-L1341.

vadz committed:
Rename formal parameter of ShowSortIndicator() to "col"

No real changes, just use a more appropriate parameter name, as it's a
column index and not just "index".

vadz committed:
Add GetUpdatedAscendingSortIndicator() helper function

It seems like this function will need to be used in every implementation
of EVT_LIST_COL_CLICK handler when using sorting, so it makes sense to
provide it in the library itself.

vadz commented:
Oops, indeed, thanks for noticing this!

I've removed it now and added a small helper, which I think would be useful in any code using sort indicators.

If you don't mind, I'm going to merge it soon because I'd like to remove the API being dropped by this PR a.s.a.p. to leave as little time as possible for people to start using it in their code.

wh11204 commented:
@vadz, would you add the helper function GetUpdatedAscendingSortIndicator() to wxHeaderCtrlSimple?

vadz commented:
@wh11204 I'm not sure, is it as useful there as here? I don't remember implementing sorting support using that control, do you have any example code doing this somewhere?

BTW, I also thought about defining some kind of moving `SortOrder` from [here](https://github.com/wxWidgets/wxWidgets/blob/b9a1931394faffc3fb95441238c82cd4e77cd15e/src/generic/datavgen.cpp#L100) to a public header and using it instead of the separate column index and ascending flag, but I'm not sure if it's worth it neither. Personally I prefer working with the objects, rather than separate ints/bools, what do the others think?

In any case, I'm going to merge this for now, and maybe improve it further later.

wh11204 commented:
Of course wxHeaderCtrlSimple does not have sort support, but your example code fragment is valid for it after creating a suitable sort function, call it p.e. MySort(): 

`void MyFrame::OnColClick(wxListEvent& event)
            {
                int col = event.GetColumn();
                if ( col == -1 )
                    return; // clicked outside any column.
                const bool ascending = header_ctrl_simple->GetUpdatedAscendingSortIndicator(col);
                MySort(MyCompareFunction, ascending);
                header_ctrl_simple->ShowSortIndicator(col, ascending);
            }`

vadz commented:
Yes, but do people actually do things like this? My impression is that `wxHeaderCtrl` would be used together with, or as a part of, some other control, which would already manage sorting internally.

Note that adding this is not completely trivial, as we'd have to add `IsAscendingSortIndicator()` to it too (and then `GetSortIndicator()` for consistency), so I'd rather not do if it's going to be really useful.

wh11204 commented:
OK
